File: phone_a_friend/backends/codex.py
"""Codex backend implementation."""

from __future__ import annotations

import shutil
import subprocess
import tempfile
from dataclasses import dataclass
from pathlib import Path


from phone_a_friend.backends import BackendError, INSTALL_HINTS


class CodexBackendError(BackendError):
    """Raised when Codex backend execution fails."""


@dataclass(frozen=True)
class CodexBackend:
    """Backend adapter for `codex exec`."""

    name: str = "codex"
    allowed_sandboxes: frozenset[str] = frozenset({"read-only", "workspace-write", "danger-full-access"})

    def run(
        self,
        *,
        prompt: str,
        repo_path: Path,
        timeout_seconds: int,
        sandbox: str,
        model: str | None,
        env: dict[str, str],
    ) -> str:
        codex_bin = shutil.which("codex")
        if not codex_bin:
            raise CodexBackendError(
                f"codex CLI not found in PATH. Install it: {INSTALL_HINTS['codex']}"
            )

        with tempfile.TemporaryDirectory(prefix="phone-a-friend-") as tmpdir:
            output_path = Path(tmpdir) / "codex-last-message.txt"
            cmd = [
                codex_bin,
                "exec",
                "-C",
                str(repo_path),
                "--skip-git-repo-check",
                "--sandbox",
                sandbox,
                "--output-last-message",
                str(output_path),
            ]
            if model:
                cmd.extend(["-m", model])
            cmd.append(prompt)

            try:
                result = subprocess.run(
                    cmd,
                    capture_output=True,
                    text=True,
                    check=False,
                    timeout=timeout_seconds,
                    env=env,
                )
            except subprocess.TimeoutExpired as exc:
                raise CodexBackendError(f"codex exec timed out after {timeout_seconds}s") from exc

            last_message = _read_output_file(output_path)
            if result.returncode != 0:
                detail = (result.stderr or result.stdout or last_message).strip()
                if not detail:
                    detail = f"codex exec exited with code {result.returncode}"
                raise CodexBackendError(detail)

            if last_message:
                return last_message

            fallback = (result.stdout or "").strip()
            if fallback:
                return fallback

        raise CodexBackendError("codex exec completed without producing feedback")


def _read_output_file(output_path: Path) -> str:
    if not output_path.exists():
        return ""
    try:
        return output_path.read_text().strip()
    except OSError as exc:
        raise CodexBackendError(f"Failed reading Codex output file: {exc}") from exc


CODEX_BACKEND = CodexBackend()

File: src/backends/codex.ts
/**
 * Codex backend implementation.
 *
 * Ported from phone_a_friend/backends/codex.py
 */

import { mkdtempSync, readFileSync, existsSync, rmSync } from 'node:fs';
import { join } from 'node:path';
import { tmpdir } from 'node:os';
import {
  BackendError,
  INSTALL_HINTS,
  isInPath,
  registerBackend,
  spawnCli,
  type Backend,
  type ReviewOptions,
  type SandboxMode,
} from './index.js';

export class CodexBackendError extends BackendError {
  constructor(message: string) {
    super(message);
    this.name = 'CodexBackendError';
  }
}

export class CodexBackend implements Backend {
  readonly name = 'codex';
  readonly localFileAccess = true;
  readonly allowedSandboxes: ReadonlySet<SandboxMode> = new Set<SandboxMode>([
    'read-only',
    'workspace-write',
    'danger-full-access',
  ]);

  async run(opts: {
    prompt: string;
    repoPath: string;
    timeoutSeconds: number;
    sandbox: SandboxMode;
    model: string | null;
    env: Record<string, string>;
  }): Promise<string> {
    if (!isInPath('codex')) {
      throw new CodexBackendError(
        `codex CLI not found in PATH. Install it: ${INSTALL_HINTS.codex}`,
      );
    }

    const tmpDir = mkdtempSync(join(tmpdir(), 'phone-a-friend-'));
    const outputPath = join(tmpDir, 'codex-last-message.txt');

    try {
      const args = [
        'exec',
        '-C',
        opts.repoPath,
        '--skip-git-repo-check',
        '--sandbox',
        opts.sandbox,
        '--output-last-message',
        outputPath,
      ];

      if (opts.model) {
        args.push('-m', opts.model);
      }

      args.push(opts.prompt);

      let stdout = '';
      try {
        const result = await spawnCli('codex', args, {
          timeoutMs: opts.timeoutSeconds * 1000,
          env: opts.env,
          label: 'codex exec',
        });
        stdout = result.stdout;
      } catch (err: unknown) {
        // On failure, check if codex wrote a useful last-message before dying
        const lastMessage = readOutputFile(outputPath);
        if (lastMessage) return lastMessage;

        if (err instanceof BackendError) {
          throw new CodexBackendError(err.message);
        }
        throw err;
      }

      // Read output file (preferred)
      const lastMessage = readOutputFile(outputPath);
      if (lastMessage) {
        return lastMessage;
      }

      // Fall back to stdout
      if (stdout) {
        return stdout;
      }

      throw new CodexBackendError('codex exec completed without producing feedback');
    } finally {
      // Clean up temp directory
      try {
        rmSync(tmpDir, { recursive: true, force: true });
      } catch {
        // Best-effort cleanup
      }
    }
  }

  async review(opts: ReviewOptions): Promise<string> {
    if (!isInPath('codex')) {
      throw new CodexBackendError(
        `codex CLI not found in PATH. Install it: ${INSTALL_HINTS.codex}`,
      );
    }

    const tmpDir = mkdtempSync(join(tmpdir(), 'phone-a-friend-'));
    const outputPath = join(tmpDir, 'codex-last-message.txt');

    try {
      const args = [
        'exec',
        'review',
        '-C',
        opts.repoPath,
        '--base',
        opts.base,
        '--sandbox',
        opts.sandbox,
        '--output-last-message',
        outputPath,
      ];

      if (opts.model) {
        args.push('-m', opts.model);
      }

      if (opts.prompt) {
        args.push(opts.prompt);
      }

      let stdout = '';
      try {
        const result = await spawnCli('codex', args, {
          timeoutMs: opts.timeoutSeconds * 1000,
          env: opts.env,
          label: 'codex exec review',
        });
        stdout = result.stdout;
      } catch (err: unknown) {
        // On failure, check if codex wrote a useful last-message before dying
        const lastMessage = readOutputFile(outputPath);
        if (lastMessage) return lastMessage;

        if (err instanceof BackendError) {
          throw new CodexBackendError(err.message);
        }
        throw err;
      }

      const lastMessage = readOutputFile(outputPath);
      if (lastMessage) {
        return lastMessage;
      }

      if (stdout) {
        return stdout;
      }

      throw new CodexBackendError('codex exec review completed without producing feedback');
    } finally {
      try {
        rmSync(tmpDir, { recursive: true, force: true });
      } catch {
        // Best-effort cleanup
      }
    }
  }
}

function readOutputFile(outputPath: string): string {
  if (!existsSync(outputPath)) {
    return '';
  }
  try {
    return readFileSync(outputPath, 'utf-8').trim();
  } catch (err) {
    throw new CodexBackendError(
      `Failed reading Codex output file: ${err}`,
    );
  }
}

export const CODEX_BACKEND = new CodexBackend();
registerBackend(CODEX_BACKEND);
